### ISSUE TITLE
Changing image_size for video segmentation

### ISSUE THREAD

**Issue by rongduan-zhu** [facebookresearch/segment-anything-2#257]
Looking at the [configs](https://github.com/facebookresearch/segment-anything-2/tree/main/sam2_configs), the default `image_size` is set to 1024. On my RTX2080, I'm getting just over 5 it/s when running inference over a video with `offload_video_to_cpu` and `async_loading_frames` set to True.

I want to see if there is a way to improve the performance and one idea I think should help is to reduce the `image_size`.

I was wondering:

1. If anyone was able to set up `SAM2VideoPredictor` and the corresponding configs so that it works with other `image_size`
2. Whether you noticed any latency reduction in inference speed

**Comment by heyoeyo** [comment]
> Whether you noticed any latency reduction in inference speed

I get the following per-frame inference times for different image sizes (on a 3090):
| Model | Time @ 1024 | Time @ 512 | Time @ 256 |
| ------|-------------| -----------| ----------|
| Large | 64ms | 17ms | 12ms |
| Base+ | 37ms | 10ms | 9ms |
| Tiny | 26ms | 8ms | 7ms |

So it's roughly a factor of 4 from 1024 down to 512, and then diminishing returns. The segmentation quality gets worse at lower sizes (as expected) and it can be hard to get a good initial prompt at the 256 sizing, so that's the main tradeoff.

> If anyone was able to set up SAM2VideoPredictor and the corresponding configs

I'm using a modified script to test the timings, but as far as I can tell, the original video predictor works as-is, as long as you pick compatible sizes (I think multiples of 128 are safe). You just have to edit the [image_size](https://github.com/facebookresearch/segment-anything-2/blob/7e1596c0b6462eb1d1ba7e1492430fed95023598/sam2_configs/sam2_hiera_b%2B.yaml#L85) entry in the .yaml file(s) and it should work. Changing the size will break the image predictor, so if you need that to still work it requires a small change (discussed in issue #138).

**Comment by rongduan-zhu** [comment]
Thank you @heyoeyo ! Changing the config worked for me. 

I previously tried to change the image size after the model is loaded in Python code and ran into an assertion error where the backbone_features size is not equal to image embedding size.

At 512, I'm able to get almost a 5x improvement (~5it/s -> ~25 it/s) with all other settings identical

**Comment by horsto** [comment]
>You just have to edit the [image_size](https://github.com/facebookresearch/segment-anything-2/blob/7e1596c0b6462eb1d1ba7e1492430fed95023598/sam2_configs/sam2_hiera_b%2B.yaml#L85) entry in the .yaml file(s) 

I tried that @heyoeyo, but no matter what I do, I get `predictor.image_size = 1024` after initialization. Any ideas for that? 

**UPDATE**: Scratch that, it was loading the wrong .yaml file in my code. Works fine!

**Comment by weirenorweiren** [comment]
> I get the following per-frame inference times for different image sizes (on a 3090):

Hi @heyoeyo, how many objects did you track for your testings? 

I tried to use SAM2 in real-time and tested with the demo from https://github.com/zdata-inc/sam2_realtime/blob/main/notebooks/realtime_detect_and_track.ipynb. I realized the FPS drops almost linearly with the number of objects to be tracked, but it might still be promising for my application with limited number of objects (at most ~50) according to the official speed measurement (https://github.com/facebookresearch/sam2?tab=readme-ov-file#sam-21-checkpoints).

**Comment by heyoeyo** [comment]
> how many objects did you track for your testings?

The times I mentioned above were for a single object. They're also from prior to the repo adding a new compilation step, which may speed things up by a factor of 2 depending on the hardware you're using.

> I realized the FPS drops almost linearly with the number of objects to be tracked

Ya it is roughly linear, something like `A + B*n` where `A` is the time needed for image encoding, `B` is the masking/memory time and `n` is the number of objects (and `A` and `B` are similarish amounts of time). For 50 objects, I'd expect it to take several seconds per frame. If you need something fast/realtime with that many objects, then something like [YOLO](https://docs.ultralytics.com/tasks/segment/) (I don't think it scales per object, from what I remember) or otherwise a [panoptic segmentation](https://paperswithcode.com/task/panoptic-segmentation) model might be better.

**Comment by weirenorweiren** [comment]
Hi @heyoeyo, thanks so much for your prompt response! 

A quick follow-up is that how much difference on FPS (or per-frame inference time) would it be expected between video and streaming settings? 

My point is that if the difference is small, considering its fantastic performance in terms of robustness, then I might think of upgrading my hardwares and then tracking less objects for my proof-of-concept. 

I appreciate any comments and thanks in advance!

**Comment by heyoeyo** [comment]
> A quick follow-up is that how much difference on FPS would it be expected between video and streaming settings?

I'm not sure I understand what you mean by the difference between video & streaming. If you mean like a live streaming source (e.g. a webcam or security camera), then that would put an upper limit on the FPS (the frame rate of the stream), and at least for tracking lots of objects, would probably require some sort of frame skipping/buffering implementation to not fall behind the live video feed. Other than that, I wouldn't expect any difference between how long the model takes to process frames from a file vs. a stream.

**Comment by weirenorweiren** [comment]
I am referring to the difference on FPS between a video file input and a live streaming source. My lab camera has an upper limit of 20FPS and I'd love the model to go far beyond 20FPS. If I remembered correctly, I saw somewhere in the Issues there are tricks for video input like batch processing to speed it up. For my case, since it's per-frame feed, batch processing seems not applied. So I am wondering that without those tricks, how much speed would be compromised? 

That reminds me to ask that for the times above, did they involve tricks for video input? If not and you happen to test the per-frame inference time by feeding video data frame by frame, I, as long as others who are interested in real-time applications, would love to see since it's a really important reference for its potential on relevant tasks!

**Comment by heyoeyo** [comment]
> for the times above, did they involve tricks for video input?

For the times I posted, they're just frame-by-frame timings. For the tables on the SAM2 page, I believe they're doing the same, except they include model compilation. The table originally showed the times without compilation, which you can view by [looking at older commits](https://github.com/facebookresearch/sam2/tree/c2ec8e14a185632b0a5d8b161928ceb50197eddc?tab=readme-ov-file#model-description).


> batch processing to speed it up

It wouldn't be possible to batch process the whole tracking calculation over time, since it's sequential (eg. the model needs the results from frame 1 before it can process frame 2, so they can't be done in parallel). It is possible to separately batch process the image encodings (over time) and at least theoretically, batch process the masking/memory steps across objects (for a single frame, not over time), though that would be more complex to implement.

For batch processing the image encoding (e.g. the `A` time in the `A + B*n` mentioned earlier), I get the following times (all using the base+ model, with 1024x1024 frame sizing and bfloat16 on a 3090, no compilation):
| Batch Size | Time per frame (ms) |
| ---------- | ------------------- |
| 1          | 20.8 |
| 8          | 19.4 |
| 16         | 18.9 |
| 32         | 19.3 |
| 60         | 19.3 |
| 64         | [error](https://github.com/pytorch/pytorch/issues/142228) |

I only left each one to run for around 7-8 seconds, so there's some variability in the numbers. In this case, it seems to be around a 10% speed up, best case, though I'd guess it's hardware dependent. Batching like this would introduce a delay, so may be problematic for real-time applications.

Batch processing the object masking/memory steps may give a bigger improvement since it's made up of smaller steps, which seems to be the main benefit of batching. However, it would be harder to organize the data for that.

**Comment by weirenorweiren** [comment]
Thanks so much for your detailed explanations and I learn more than I expect! Since I don't have too much background in AI, it took me some time to digest. Now I want to clarify a few things and make sure I understand it correctly. Please bear with me! 

1. Regarding the batch processing of image encodings, I could start encoding the next frame before the encoding of the current frame is done. There is a delay because several subsequent frames has been passed before I could get the encoding from the first frame. Is it right? 
2. What does the batch size refer to? Does it imply the maximum number of frames that are processed simultaneously by a given GPU?
4. Regarding the batch processing of the masking/memory steps, I think of two ways. One is similar to the batch processing of image encodings that I process all masks at the current frame simultaneously. Another is to merge all masks into a new mask at the current frame and keep doing single object tracking. Do they make sense to you? What's your thoughts on them?
5. For batch processing of either the image encodings or the masking/memory steps, how does a GPU help? My intuitive understanding is _each frame within the batch during the image encodings_ and _masks for each object within a frame during the masking/memory steps_ are allocated to a different part of GPU for parallel computing. Does it make sense? If so, does the GPU manage to start processing a new batch for the image encodings or a new frame for the masking/memory steps (e.g., When for the current frame the `mask_for_object_1` is done and the `mask_for_object_2` is being generated, `mask_for_object_1` is being generated for the next frame.) before the current batch or frame is done with processing? 
7. Would model compilation be applied for real-time/live-streaming setting that I am interested in? 
8. What's the relationship between compilation and batch processing? In the README, there is `compile_image_encoder: True` and I feel it's equivalent to the batch processing of image encodings. If that's the case, then I feel the model compilation sounds like a batch processing and I intuitively see it the same as batch processing of the masking/memory steps. However, it appears not the case as I read your latest comments. I appreciate any clarifications on this!

**Comment by heyoeyo** [comment]
I think answering these out of order might make the most sense:

> 2. What does the batch size refer to?

Just as you say, it's a way of parallel processing the frames. The main reason to do it is that it reduces the amount of cpu-to-gpu communication, which takes time. It's like going to the store and buying a 'batch' of 5 oranges, vs. going to the store 5 times and buying one orange each time. In both cases you're doing the work of carrying 5 oranges, but you save a lot of travel overhead by getting them all at once. The downside is you need room to carry the oranges, maybe you can't handle a batch of 1000. By analogy, you need more VRAM for larger batches on the GPU.

> 1. ... There is a delay because several subsequent frames has been passed before I could get the encoding from the first frame.

Sorry I should have clarified. The delay in batching is because you need to wait for all the frames to form the batch. For example, if you use a batch size of 60, you would need to wait for 60 frames before giving the data to the GPU. Even if the GPU finished instantly, the results for the first-most frame of the batch would have been delayed by waiting for the 60th frame to arrive to begin processing (ex. at 20FPS, it would introduce a 3 second delay).
The delay you mentioned is also true, a bigger batch will take longer so you don't get the results as quickly. For the times I mentioned above, you'd only wait 20.8ms for encoding a single frame, whereas a batch of 16 frames takes 302ms (16 * 18.9ms/frame).

> 6. What's the relationship between compilation and batch processing?

Compilation is for converting the model code (given in python) into a more efficient form (more info on the [pytorch site](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html)). There's maybe some analogy with batching, but mostly it's a separate idea (it has to do with code/instructions, rather than data).

> 5. Would model compilation be applied for real-time/live-streaming setting that I am interested in?

Compilation can be enabled with the [vos_optimized](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/build_sam.py#L107) parameter when creating the video predictor.
Or you can manually run the [compilation steps](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/sam2_video_predictor.py#L983) before using the model.

> 4. ... does the GPU manage to start processing a new batch before the current batch is done with processing?

I think this is about pipelining/staggering the work being done. For most (all?) current GPUs, the image encoding/masking steps will sort of 'max out' the hardware. If you try to queue image encodings for frame 2 while masks for frame 1 are still being generated, it should work if there's enough VRAM, but both tasks will be slowed down so that the total time taken is similar to if they were run sequentially (probably hardware dependent). Coordinating that sort of data processing would be tricky, so a simpler way to test this might be to launch two copies of the video processing and measure the total time, vs. running one instance twice.
For example, when I run the base+ model on a video, it takes around 37ms per frame. If I launch a second instance, the times are less consistent but both take around 70-80ms per frame.

> 3. Another is to merge all masks into a new mask at the current frame and keep doing single object tracking

This can provide a huge speed up if you have a video where it works. Generally the SAM models won't do well tracking different objects as a single mask, but if the objects stay close and are similar looking, then it does tend to work to some extent.

**Comment by weirenorweiren** [comment]
Thanks @heyoeyo - GPT for your lecture and it really helps! 

One quick follow-up is that could you elaborate the below statement? I have 2 types of targets and each target has 10+ in number. One looks like a solid circle and the other looks like a ring/donut. The tricky part is that the solid circle could have a significant overlapping with the feature of the ring/donut. 
> but if the objects stay close and are similar looking, then it does tend to work to some extent

**Comment by heyoeyo** [comment]
> One quick follow-up is that could you elaborate the below statement?

From what I've seen, the model can only track multiple objects using 1 mask when there's a lot of similarity between objects. As an example, you can get the model to segment an entire forest, or you could get it to segment individual trees, and tracking works in both cases, but is significantly faster if you track the entire forest as one object. However, if the objects are moving apart/independently of one another or changing appearance, then the tracking can start to struggle.

> The tricky part is that the solid circle could have a significant overlapping with the feature of the ring/donut

If the masking tends to interchange rings/solid circles while tracking, then it might be worth doing some kind of post-processing of the masks. For example, if you're tracking the solid circles, then checking that the mask predictions don't have holes could help prevent the tracking from switching from circles to ring objects (or vice versa, e.g. make sure ring masks do have holes). If the model does predict a ring mask when it's supposed to be a solid circle, you could skip that frame of tracking or otherwise check the other mask predictions (the model always generates 4 predictions. The mask [selection happens here](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/modeling/sam2_base.py#L382-L386)), since one of them could be the solid circle.
